source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -474,28 +467,23 @@ bool Filter::maybeTunnel(const std::string& cluster_name) {
                        "http2_protocol_options on the cluster.");
       return false;
     }
-    Http::ConnectionPool::Instance* conn_pool = cluster_manager_.httpConnPoolForCluster(
-        cluster_name, Upstream::ResourcePriority::Default, absl::nullopt, this);
-    if (conn_pool) {
-      upstream_ = std::make_unique<HttpUpstream>(*upstream_callbacks_,
-                                                 config_->tunnelingConfig()->hostname());
-      HttpUpstream* http_upstream = static_cast<HttpUpstream*>(upstream_.get());
-      Http::ConnectionPool::Cancellable* cancellable =
-          conn_pool->newStream(http_upstream->responseDecoder(), *this);
-      if (cancellable) {
-        ASSERT(upstream_handle_.get() == nullptr);
-        upstream_handle_ = std::make_shared<HttpConnectionHandle>(cancellable);
-      }
+
+    generic_conn_pool_ = std::make_unique<HttpConnPool>(cluster_name, cluster_manager_, this,
+                                                        config_->tunnelingConfig()->hostname(),
+                                                        *upstream_callbacks_);
+    if (generic_conn_pool_->valid()) {
```

Comment:
Alternative idea (which may or may not be better): instead of having a `valid()` method, have `newStream()` immediately call `onGenericPoolFailure()` in the case where it isn't valid.

Reply:
The validity check is going to be hidden away in a follow-up PR: if there's no conn pool, the factory will just return nullptr as the HTTP variant does:

https://github.com/envoyproxy/envoy/blob/master/source/extensions/upstreams/http/http/config.cc#L16
https://github.com/envoyproxy/envoy/blob/master/source/common/router/router.cc#L614
https://github.com/envoyproxy/envoy/blob/master/source/common/router/router.cc#L1540